### PR TITLE
fix: do not deregister GitLab tokens which are still in use

### LINKED ISSUE #1062: 'remove-gitlab-registration' script can invalidate in-use runner authentication token
# Describe the bug

We're in the process of upgrading to `7.2.2` + the [new runner registration workflow](https://docs.gitlab.com/ee/ci/runners/new_creation_workflow.html), and discovered an unfortunate race condition with `remove-gitlab-registration`.

Sometimes when cycling the agent instance, the AWS ASG will spin up the new instance before tearing down the old one. In that case, the following sequence of events can occur:

| Old Instance | New Instance |
| ------------- | ------------- |
| _starts_ | _n/a_ |
| [obtains `Runner Token 1` from Gitlab](https://github.com/cattle-ops/terraform-aws-gitlab-runner/blob/8b73d42fe21602a343155743cb5d4beea0733b89/template/gitlab-runner.tftpl#L32-L81) | _n/a_ |
| [adds `Runner Token 1` to SSM](https://github.com/cattle-ops/terraform-aws-gitlab-runner/blob/8b73d42fe21602a343155743cb5d4beea0733b89/template/gitlab-runner.tftpl#L82) | _n/a_ |
| _running_ | _n/a_ |
| _running_  | _starts_ |
| _running_ | [fetches and validates `Runner Token 1` from SSM](https://github.com/cattle-ops/terraform-aws-gitlab-runner/blob/8b73d42fe21602a343155743cb5d4beea0733b89/template/gitlab-runner.tftpl#L21-L28) |
| _running_ | _finishes startup_ |
| _begins shutdown_ | _running_ |
| [remove_gitlab_registration.sh](https://github.com/cattle-ops/terraform-aws-gitlab-runner/blob/8b73d42fe21602a343155743cb5d4beea0733b89/template/gitlab-runner.tftpl#L174-L180) runs and invalidates `Runner Token 1` | _running_ |
| _finishes shutdown_ | _receives 403 from Gitlab because `Runner Token 1` is now invalid_ |

# To Reproduce

The exact steps to reproduce seem to be tricky because the ASG seems inconsistent in how it's cycling the instances, but I believe it's:
1. Upgrade module to `7.2.2`
2. Use the [new runner registration workflow](https://docs.gitlab.com/ee/ci/runners/new_creation_workflow.html) in the [gitlab-runner.tftpl](https://github.com/cattle-ops/terraform-aws-gitlab-runner/blob/8b73d42fe21602a343155743cb5d4beea0733b89/template/gitlab-runner.tftpl#L33-L62) script
    * To enable this, we're passing in:
    * `runner_gitlab.access_token_secure_parameter_store_name`
    * `runner_gitlab_registration_config.type`
    * `runner_gitlab_registration_config.project_id`
3. Run `terraform apply`
4. Wait for the ASG to stabilize
5. Make a change to the agent's user-data (e.g. by modifying `runner_install.pre_install_script`), and re-run `terraform apply`
6. Wait for the ASG to stabilize
7. Verify that the new agent instance cannot contact Gitlab because of `Forbidden` errors

# Expected behavior

The new agent instance can continue to run jobs from Gitlab.

# Additional context

As a concrete example, the new instance fetched and validated the token at `Jan 05 18:49:04 `, but the the original instance then started shutdown at `Jan 05 18:49:17` and invalidate the token, after which the new instance started logging the following:
```
Jan 05 18:50:51 gitlab-runner[2957]: {"level":"error","msg":"Checking for jobs... forbidden","runner":"<redacted>","status":"POST https://gitlab.com/api/v4/jobs/request: 403 Forbidden","time":"2024-01-05T18:50:51Z"}
Jan 05 18:50:54 gitlab-runner[2957]: {"level":"error","msg":"Checking for jobs... forbidden","runner":"<redacted>","status":"POST https://gitlab.com/api/v4/jobs/request: 403 Forbidden","time":"2024-01-05T18:50:54Z"}
Jan 05 18:50:57 gitlab-runner[2957]: {"level":"error","msg":"Checking for jobs... forbidden","runner":"<redacted>","status":"POST https://gitlab.com/api/v4/jobs/request: 403 Forbidden","time":"2024-01-05T18:50:57Z"}
Jan 05 18:50:57 gitlab-runner[2957]: {"level":"error","msg":"Runner \"<redacted>\" is unhealthy and will be disabled for 1h0m0s seconds!","time":"2024-01-05T18:50:57Z","unhealthy_requests":3,"unhealthy_requests_limit":3}
```

---

As a temporary workaround, we're passing the following into `runner_install.start_script`:
```sh
# This can be removed once https://github.com/cattle-ops/terraform-aws-gitlab-runner/issues/1062 is resolved.
echo 'Disabling "Remove the GitLab Runner from GitLab at shutdown" job...'
chmod a-x /opt/remove_gitlab_registration.sh || true
systemctl disable remove-gitlab-registration.service || true
systemctl stop remove-gitlab-registration.service || true
echo 'Done!'
```

### PR DESCRIPTION
**Pull request by kayman-mk**
## Description

So far we haven't tracked the usage of the GitLab Runner token. In case the Runner is stopped and a new one is spawned, race conditions might occur as explained in #1062. In consequence the new Runner used a token deleted by the old Runner. This leads into downtimes as the token is no longer valid and can't be used.

This PR converts the token into a JSON format and ensures that tokens which are still in use, are not deregistered.

Fixes #1062 

## Migrations needed

In case you want to rollback to a previous version you have to convert the SSM parameter containing the token and usage counter in JSON format back to a plain token string.

## Verification

- [x] script changes were tested locally
- [x] Test Runner was started to ensure that the token conversion works

### CODE REVIEW

File: template/gitlab-runner.tftpl [hunk 1]
```diff
@@ -183,10 +197,38 @@ EOF
...
+
+if [[ $deregister_runner == "true" ]]; then
+  echo "Removing Gitlab Runner ..."
 
+  aws ssm put-parameter --overwrite --type SecureString  --name "${secure_parameter_store_runner_token_key}" --region "${secure_parameter_store_region}" --value="{\"token\": \"null\", \"in_use_counter\": 0}" 2>&1
+  curl -sS ${curl_cacert} --request DELETE "${runners_gitlab_url}/api/v4/runners" --form "token=$token" 2>&1
+else
+  usage_counter=$(echo $json_token | jq -r .usage_counter)
+  usage_counter=$(($usage_counter-1))
+  json_token=$(echo $json_token | jq ".usage_counter = $usage_count er")
```
Comment (long-wan-ep) [inline template/gitlab-runner.tftpl:226]: I tried this jq expression locally but couldn't get it to work. I'm not familiar with the `er` part, what does it do?

**Review by long-wan-ep — COMMENTED** [review]
I think it still needs a section for incrementing the counter, but looking good so far.

File: template/gitlab-runner.tftpl (hunk 1 above)
Comment (kayman-mk) [reply template/gitlab-runner.tftpl:226]: Well, it does nothing. Simply wrong. Thanks for pointing this out. Curious, I tested it on a Runner instance. Maybe the space came in by accident.

Comment (long-wan-ep) [reply template/gitlab-runner.tftpl]: Thanks for fixing.

**Comment by kayman-mk** [comment]
> I think it still needs a section for incrementing the counter, but looking good so far.

@long-wan-ep Yes, missed that and added it a second ago.

**Comment by long-wan-ep** [comment]
@kayman-mk Thanks for adding that. I tested this out by deploying a test runner and triggering instance refreshes on the ASG. The increment is working, but the decrement is not. I took a look inside the runner instance's `/opt/remove_gitlab_registration.sh` and the script had all of its variables expanded, this is what it looks like:
```
#!/bin/bash
json_token={
  "token": "xxxxxxxxxxxxxx",
  "usage_counter": 2
}
deregister_runner=true

# for  module versions >7.4.1 the token is JSON
if [[ "{
  "token": "xxxxxxxxxxxxxx",
  "usage_counter": 2
}" =~ ^\{.* ]]; then
  usage_counter=2

  # ensure that the token is not in use by another Runner
  if [[ 2 -gt 1 ]]; then
    deregister_runner=false
    token="not needed"
  else
    token=xxxxxxxxxxxxxx
  fi
else
  token={
  "token": "xxxxxxxxxxxxxx",
  "usage_counter": 2
}
fi

if [[  == "true" ]]; then
  echo "Removing Gitlab Runner ..."

  aws ssm put-parameter --overwrite --type SecureString  --name "runner-usage-counter-test-runner-token" --region "us-west-2" --value="{\"token\": \"null\", \"in_use_counter\": 0}" 2>&1
  curl -sS  --request DELETE "https://gitlab.com//api/v4/runners" --form "token=xxxxxxxxxxxxxx" 2>&1
else
  usage_counter=2
  usage_counter=1
  json_token={
  "token": "xxxxxxxxxxxxxx",
  "usage_counter": 2
}

  aws ssm put-parameter --overwrite --type SecureString  --name "runner-usage-counter-test-runner-token" --region "us-west-2" --value="{
  "token": "xxxxxxxxxxxxxx",
  "usage_counter": 2
}" 2>&1

  echo "Token still in use. GitLab Runner not removed from GitLab."
fi
```

File: template/gitlab-runner.tftpl [hunk 2]
```diff
@@ -181,11 +202,35 @@ WantedBy=multi-user.target
...
+  deregister_runner=false
+  token="not needed"
+else
+  token=$(echo $json_token | jq -r .token)
+fi
+
+if [[ $deregister_runner == "true" ]]; then
+  echo "Removing Gitlab Runner ..."
+
+  aws ssm put-parameter --overwrite --type SecureString  --name "${secure_parameter_store_runner_token_key}" --region "${secure_parameter_store_region}" --value="{\"token\": \"null\", \"in_use_counter\": 0}" 2>&1
```
Comment (long-wan-ep) [inline template/gitlab-runner.tftpl:223]:
```suggestion
  aws ssm put-parameter --overwrite --type SecureString  --name "${secure_parameter_store_runner_token_key}" --region "${secure_parameter_store_region}" --value="{\"token\": \"null\", \"usage_counter\": 0}" 2>&1
```

**Review by long-wan-ep — CHANGES_REQUESTED** [review]
Overall looks good, one minor thing.

File: template/gitlab-runner.tftpl (hunk 2 above)
Comment (kayman-mk) [reply template/gitlab-runner.tftpl]: The only thing I didn't test. Good job!

**Pull request merged**